internal/services/v1/experimental.go

```diff
@@ -386,6 +406,153 @@ func (es *experimentalServer) BulkExportRelationships(
 	return nil
 }
 
+func (es *experimentalServer) BulkCheckPermission(ctx context.Context, req *v1.BulkCheckPermissionRequest) (*v1.BulkCheckPermissionResponse, error) {
+	atRevision, checkedAt, err := consistency.RevisionFromContext(ctx)
+	if err != nil {
+		return nil, es.rewriteError(ctx, err)
+	}
+
+	// Identify checks with same permission+subject over different resources and group them. This is doable because
+	// the dispatching system already internally supports this kind of batching for performance.
+	groupingParams := groupingParameters{
+		atRevision:           atRevision,
+		maxCaveatContextSize: es.maxCaveatContextSize,
+		maximumAPIDepth:      es.maximumAPIDepth,
+	}
+	groupedItems, err := groupItems(ctx, groupingParams, req.Items)
+	if err != nil {
+		return nil, es.rewriteError(ctx, err)
+	}
+
+	bulkResponseMutex := sync.Mutex{}
+	resp := &v1.BulkCheckPermissionResponse{CheckedAt: checkedAt}
+	tr := taskrunner.NewPreloadedTaskRunner(ctx, es.bulkCheckMaxConcurrency, len(groupedItems))
+
+	appendResultsForError := func(group groupedCheckParameters, err error) error {
+		bulkResponseMutex.Lock()
+		defer bulkResponseMutex.Unlock()
```

Comment:
We can move this to just above the for loop, because rewriting the error doesn't require holding the lock.